Volunteer Application Form

JessSumner opened the pull request:
Why:
The rescue has an applciation form for all volunteers.

This PR:
Adds a model for volunteer_application to handle the data.
Adds a controller for volunteer_applications.
Adds a view which renders the volunteer application form fields.
Extracts the styling and error views into a separate file as they are used by both the adoption application form and the volunteer application form.

JessSumner committed:
Foster Application form built

FosterApplication model generated
Validations added to the model for valid phone numbers and emails.
Routes set up and added to the volunteer page.
Form fields set up.
Dependent fields created

JessSumner committed:
Corrected a few typos

In the controller the save method was being called on @adoption_application. This has been corrected to @foster_application.
The model was calling a validation on :cell rather than :phone. This has been ammended.

JessSumner committed:
Moved error messages partial

The adoptions application form has it's own partial for rendering error messages. I will need one for the foster application too.

Instead of duplicating the file I have modified the error partial to call the errors on an object rather than specifically the @adoption_application. The object is passed in when the partial is rendered. The partial has been moved to a separate folder in views called error_messages. This way the same partial can be rendered if needed by the foster application form.

JessSumner committed:
Updated form labels to include a question.

Some fields required a more detailed question than the one generated.

More detailed questions have been added.

JessSumner committed:
Error messages work for foster application

Due to the neutral location of the error messages partial. It is simple and clear how to incorporate the same error messages into the foster application.

JessSumner committed:
Updated Foster Application to Volunteer Application

Why:
This application form covers the choice of fostering or being a quarantine home.
Need to make sure that applicants can commit to adoption appointments.

This commit:
Application has been renamed to Volunteer Application Form. This is a more descriptive title.
Renamed some data fields to be consistent with the Adoption Application Form.
Added a data columns/radio buttons/notes for adoption_appointments. This is to ensure that applicants are aware of the commitment of meeting with adopters for their rats.
Added an any questions and comments section so that an applicant can add extra information if desired.
Changed the work_location question to a radio button.

JessSumner committed:
Organized stylesheets

Why:
Styles for adoption application also apply to the volunteer application form.

This commit:
Pulled all the styling for the form fields into a form_fields.css.scss file.
Pulled all the styling for the error messages into an errors.css.scss file.
Added some line breaks to the end of the radio buttons so that they are aligned correctly.

JessSumner committed:
Hound corrections

Deleted test files as they are not in use.
In errors stylesheet: changed id selector to a class selector, replaced underscores with hyphens for field with errors selector. Changes were then reflected in the error_messages views.
In models, reduced line to under 80 characters.
Added flash.now to the controller as the flash messages were remaining after a page move.

JessSumner committed:
Hound corrections - whitespace and indentations

Trailing whitespace removed in models.
Ammended indentation to be consistent in controller.

JessSumner commented:
@MattMSumner @kenyonj @halogenandtoast I'm hoping this is a relatively simple PR. It is similar to the adoption application form which I has already been merged. If you have time I would appreciate any comments you have on it. Thank you.

halogenandtoast commented on app/assets/javascripts/foster_applications.coffee:
```diff
@@ -0,0 +1,3 @@
+# Place all the behaviors and hooks related to the matching controller here.
```
You can probably delete this file since it is empty.

halogenandtoast commented on app/controllers/adoption_applications_controller.rb:
```diff
@@ -6,11 +6,11 @@ def new
     @adoption_application = AdoptionApplication.new(adoption_application_params)
     if @adoption_application.save
-      flash[:success] =
+      flash.now[:success] =
```
This does not need to use the `.now` function since you redirect. The one below does because you render.

halogenandtoast commented on app/controllers/volunteer_applications_controller.rb:
```diff
@@ -0,0 +1,49 @@
+    @volunteer_application = VolunteerApplication.new
+                             (volunteer_application_params)
+    if @volunteer_application.save
+      flash.now[:success] =
```
This doesn't need to use `.now` because you redirect. The one below does because you call `render`

halogenandtoast commented on app/helpers/foster_applications_helper.rb:
```diff
@@ -0,0 +1,2 @@
+module FosterApplicationsHelper
```
You can delete this file since it's empty.

halogenandtoast commented on app/views/adoption_applications/new.html.erb:
```diff
@@ -1,7 +1,7 @@
 
 <% if @adoption_application.errors.any? %>
-  <%= render "error_messages" %>
+  <%= render 'error_messages/error_messages.html.erb', object: @adoption_application %>
```
You can leave off the `.html.erb`. I'd also use double quotes because it's consistent with what you're doing elsewhere.

halogenandtoast commented on test/fixtures/foster_applications.yml:
```diff
@@ -0,0 +1,11 @@
+# Read about fixtures at http://api.rubyonrails.org/classes/ActiveRecord/FixtureSet.html
```
I'd delete this file.

JessSumner committed:
Removed redundant fields made a couple of corrections

Ammended the flash messages. If using redirect there is no need to use flash.now. However if rendering a template flash.now is useful. Reflected this in the adoption application controller and the volunteer application folder.
Removed test/fixtures file, coffee file and helper file for the foster application as none of these are being used.
When rendering a file it is not necessary to put the full file extension. Just the file name and the folder path to it relative to the file which is rendering. The file extension has been deleted from the render call in adoption application view. Also switched to double quotes as that is my style pattern.

halogenandtoast commented on app/models/volunteer_application.rb:
```diff
@@ -0,0 +1,21 @@
+class VolunteerApplication < ActiveRecord::Base
+  VALID_EMAIL_REGEX = /\A[\w+\-.]+@[a-z\d\-]+(\.[a-z\d\-]+)*\.[a-z]+\z/i
```
Instead of doing this I'd consider using a gem like https://github.com/balexand/email_validator

Email regexes can be quite complicated and I think being able to say:

``` ruby
validates :email, email: true, presence: true
```

is nice and keeps your model a bit cleaner.

halogenandtoast commented on app/models/volunteer_application.rb:
```diff
@@ -0,0 +1,21 @@
+class VolunteerApplication < ActiveRecord::Base
+  VALID_EMAIL_REGEX = /\A[\w+\-.]+@[a-z\d\-]+(\.[a-z\d\-]+)*\.[a-z]+\z/i
+  VALID_PHONE_REGEX = /[0-9]{10}/
```
I'd consider pulling this out into it's own validator. It's not necessary per say, but I think it could be good practice and it will keep this model leaner. If you have a custom validator you should be able to do

``` ruby
validates :cell, phone: true, presence: true
```

Check out http://guides.rubyonrails.org/active_record_validations.html#performing-custom-validations to see how rails suggests doing this. If you decide to create a custom validator, I'd suggest throwing it into `app/validators`

halogenandtoast replied:
Also you'll be able to use this and the email validator in the `AdoptionApplication` model

kenyonj commented on app/views/volunteers/show.html.erb:
```diff
@@ -38,7 +38,7 @@
     </p>
     <p>
-      If you are interested in either fostering or being a quarantine home please fill out our <%=link_to "Foster Application", '#' %> and we will be in touch shortly.
+      If you are interested in either fostering or being a quarantine home please fill out our <%=link_to "Volunteer Application", new_volunteer_application_path %> and we will be in touch shortly.
```
space missing after `<%=` , also thoughts on breaking this line up?

halogenandtoast commented on app/models/volunteer_application.rb:
```diff
@@ -0,0 +1,21 @@
+class VolunteerApplication < ActiveRecord::Base
+  VALID_EMAIL_REGEX = /\A[\w+\-.]+@[a-z\d\-]+(\.[a-z\d\-]+)*\.[a-z]+\z/i
+  VALID_PHONE_REGEX = /[0-9]{10}/
+  before_save :downcase_email
```
You don't need to change this, but I wanted to warn you about the dangers of `before_save`. It's often the source of unexpected behavior because it's very difficult to save without `before_save`s triggering. I tend to never use `before_save` unless I have to. I think it's fine here, but I'd always suggest trying to do something else first before adding one. Same with `before_create`, `before_validation`, etc.

JessSumner committed:
Included email_validator gem

Email regex is complicated and therefore delegating this logic hopefully mitigates any potential errors.
Removed the regex from the model.
Ammended the controller for the volunteer application. Moving part of the method to another line meant that rendering new after errors did not save the values entered. This has been corrected.

JessSumner committed:
Validations and line length corrections

Pulled out the cell phone validation into it's own file. This allows it to be used for both models but only defined once.

Added a method in config/application that when field_with_errors is called it creates a new proc called field-with-errors. This allows me to ammend the css class to comply with my style guide.

Shortened several lines in adoption_application_controller, volunteer_application_controller and the volunteers/show view to be under 80 characters.

JessSumner committed:
Style guide corrections

Corrections made to the config/application file. %Q changed to % Proc.new changed to proc and instance changed to _instance.
Line shortened to under 80 characters in cell validation file.
